src/cpp/rtps/reader/StatefulReader.cpp

```diff
@@ -641,7 +641,15 @@ bool StatefulReader::processDataMsg(
                 }
                 datasharing_pool->get_datasharing_change(change->serializedPayload, *change_to_add);
             }
-            else if (!payload_pool_->get_payload(change->serializedPayload, change_to_add->serializedPayload))
+            else if (payload_pool_->get_payload(change->serializedPayload, change_to_add->serializedPayload))
+            {
+                if (change->serializedPayload.payload_owner == nullptr)
+                {
+                    change_to_add->serializedPayload.payload_owner->get_payload(change_to_add->serializedPayload,
+                            change->serializedPayload);
```

Comment:
We can try the following:

```suggestion
                    payload_pool_->get_payload(change_to_add->serializedPayload, change->serializedPayload);
```